oak/server/rust/oak_runtime/src/node/http/mod.rs

```diff
@@ -0,0 +1,520 @@
+        if let Err(err) = runtime.channel_close(startup_handle) {
+            error!(
+                "Failed to close initial inbound channel {}: {:?}",
+                startup_handle, err
+            );
+        }
+
+        // Build a service to process all incoming HTTP requests.
+        let generic_handler = HttpRequestHandler {
+            runtime,
+            invocation_channel,
+        };
+        let server = self.make_server(generic_handler, notify_receiver);
+
+        // Create an Async runtime for executing futures.
+        // https://docs.rs/tokio/
+        let mut async_runtime = create_async_runtime();
+
+        // Start the HTTP server.
+        info!(
+            "{}: Starting HTTP server pseudo-Node on: {}",
+            self.node_name, self.address
+        );
+        async_runtime.block_on(server);
+    }
+}
+
+fn create_async_runtime() -> tokio::runtime::Runtime {
+    tokio::runtime::Builder::new()
+        // Use simple scheduler that runs all tasks on the current-thread.
+        // https://docs.rs/tokio/0.2.16/tokio/runtime/index.html#basic-scheduler
+        .basic_scheduler()
+        // Enables the I/O driver.
+        // Necessary for using net, process, signal, and I/O types on the Tokio runtime.
+        .enable_io()
+        // Enables the time driver.
+        // Necessary for creating a Tokio Runtime.
+        .enable_time()
+        .build()
+        .expect("Couldn't create Async runtime")
+}
+
+/// [`HttpRequestHandler`] handles HTTP requests from a client and sends HTTP responses back.
+#[derive(Clone)]
+struct HttpRequestHandler {
+    /// Reference to the Runtime in the context of this HTTP server pseudo-Node.
+    runtime: RuntimeProxy,
+    /// Channel handle used for writing HTTP invocations.
+    invocation_channel: oak_abi::Handle,
+}
+
+impl HttpRequestHandler {
+    fn handle(
+        &self,
+        request: HttpRequest,
+    ) -> Pin<Box<dyn Future<Output = Result<Response<Body>, OakStatus>> + Send + Sync>> {
+        let handler = self.clone();
+        let future = async move {
+            let oak_label = get_oak_label(&request)?;
+            info!(
+                "handling HTTP/2 request; request size: {} bytes, label: {:?}",
+                request.body.len(),
+                oak_label
+            );
+
+            debug!("inject the request into the Oak Node");
+            let response = handler
+                .inject_http_request(request, &oak_label)
+                .map_err(|_| OakStatus::ErrInternal)?;
+
+            Ok(response.to_response())
+        };
+
+        Box::pin(future)
+    }
+
+    fn inject_http_request(
+        &self,
+        request: HttpRequest,
+        label: &Label,
+    ) -> Result<HttpResponseIterator, ()> {
+        // Create a pair of temporary channels to pass the HTTP/2 request and to receive the
+        // response.
+        let pipe = Pipe::new(&self.runtime.clone(), label)?;
+
+        // Serialize HTTP/2 request into a message.
```

Comment:
I am not sure whether the `HTTP/2` vs `HTTP` terminology is something I should pay attention throughout, or if it is being used interchangeably. If it's not intentional, I think sticking to `HTTP` would probably be more straightforward.